schema/api.go

```diff
@@ -0,0 +1,98 @@
+// existing input or output fields.
+// Input fields also cannot reference any unsealed structs, directly or transitively,
+// because these types allow adding new fields.
+// Adding new input fields to a method introduces the possibility that a newer client
+// will send an incomprehensible message to an older server.
+// The only safe ways that input field schemas can be extended are by adding
+// new values to EnumType's and new cases to OneOfType's.
+// It is a COMPATIBLE change to add new methods to an API and to add new output fields
+// to existing methods.
+// Output fields can reference any sealed or unsealed StructType, directly or transitively.
+//
+// Existing protobuf APIs could also be mapped into APIDefinitions, and used in the following ways:
+// - to produce, user-friendly deterministic JSON
+// - to produce a deterministic binary encoding
+// - to check for compatibility in a way that is more appropriate to blockchain applications
+// - to use any code generators designed to support this spec as an alternative to protobuf
+// Also, a standardized way of serializing schema types as protobuf could be defined which
+// maps to the original protobuf encoding, so that schemas can be used as an interop
+// layer between different less expressive encoding systems.
+//
+// Existing EVM contract APIs expressed in Solidity could be mapped into APIDefinitions, and
+// a mapping of all schema values to ABI encoding could be defined which preserves the
+// original ABI encoding.
+//
+// In this way, we can define an interop layer between contracts in the EVM world,
+// SDK modules accepting protobuf types, and any API using this schema system natively.
+type APIDescriptor struct {
+	// Name is the versioned name of the API.
+	Name string
+
+	// Methods is the list of methods in the API.
+	// It is a COMPATIBLE change to add new methods to an API.
+	// If a newer client tries to call a method that an older server does not recognize it,
+	// an error will simply be returned.
+	Methods []MethodDescriptor
+}
+
+// MethodDescriptor describes a method in the API.
+type MethodDescriptor struct {
+	// Name is the name of the method.
+	Name string
+
+	// InputFields is the list of input fields for the method.
+	//
+	// It is an INCOMPATIBLE change to add, remove or update input fields to a method.
+	// The addition of new fields introduces the possibility that a newer client
+	// will send an incomprehensible message to an older server.
+	// InputFields can only reference sealed StructTypes, either directly and transitively.
+	//
+	// As a special case to represent protobuf service definitions, there can be a single
+	// unnamed struct input field that code generators can choose to either reference
+	// as a named struct or to expand inline as function arguments.
+	InputFields []Field
+
+	// OutputFields is the list of output fields for the method.
+	//
+	// It is a COMPATIBLE change to add new output fields to a method,
```

Comment:
I think for now we should amend this until we have strong guarantees that it cannot be consensus breaking.